### PR TITLE
Add the public header Version.h

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR adds the public header `Version.h` which contains:

* the traditional version macros `PRECICE_VERSION_MAJOR`, `PRECICE_VERSION_MINOR`, `PRECICE_VERSION_PATCH`
* the version as a string literal `PRECICE_VERSION`
* a macro to check if the preCICE version is greater equal than the provided version `PRECICE_VERSION_GTE(major,minor,patch)`

This header is C-compatible (hence `.h`) and included by `SolverInterface.hpp` as well as `SolverInterfaceC.h`, so these macros are available in user-code without any further changes.

## Motivation and additional information

This allows applications to differentiate between various preCICE versions without the need to fetch this information from a package manager or the like.

## Author's checklist

* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I ran `make format` to ensure everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [x] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -327,6 +327,7 @@ endif()
 include(GenerateVersionInformation)
 include(${CMAKE_CURRENT_LIST_DIR}/cmake/DetectGitRevision.cmake)
 configure_file("${PROJECT_SOURCE_DIR}/src/precice/impl/versions.hpp.in" "${PROJECT_BINARY_DIR}/src/precice/impl/versions.hpp" @ONLY)
+configure_file("${PROJECT_SOURCE_DIR}/src/precice/Version.h.in" "${PROJECT_BINARY_DIR}/src/precice/Version.h" @ONLY)
```
Comment (DavidSCN) [inline CMakeLists.txt:330]: Why do we have `Version.h` starting with a capital `V` and the `versions.hpp` starting with a small `v`?

File: docs/changelog/1392.md [hunk 2]
```diff
@@ -0,0 +1 @@
+- Added a new interface file `precice/Version.h`, which provides version macros `PRECICE_VERSION_MAJOR`, `PRECICE_VERSION_MINOR`, `PRECICE_VERSION_PATCH` as well as the stringified version `PRECICE_VERSION`. This file is now included by `SolverInterface.hpp` and `SolverInterfaceC.h`.
```
Comment (DavidSCN) [inline docs/changelog/1392.md:1]: From the user side, the version macro `PRECICE_VERSION_GTE´ should be of practical relevance. We should mention it in the changelog.

File: src/precice/SolverInterface.hpp [hunk 3]
```diff
@@ -4,6 +4,7 @@
 #include <set>
 #include <string>
 #include <vector>
+#include "precice/Version.h"
```
Comment (DavidSCN) [inline src/precice/SolverInterface.hpp:7]: Do we want to include it by default? Your changelog entry sounds more like the user has to include the file additionally.

File: src/precice/Version.h.in [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+#define PRECICE_VERSION_MAJOR @preCICE_VERSION_MAJOR@
+#define PRECICE_VERSION_MINOR @preCICE_VERSION_MINOR@
+#define PRECICE_VERSION_PATCH @preCICE_VERSION_PATCH@
+
+#define PRECICE_VERSION "@preCICE_VERSION@"
+
+// clang-format on
+
+#define PRECICE_VERSION_GTE(major, minor, patch) (                       \
```
Comment (DavidSCN) [inline src/precice/Version.h.in:13]: Forwarding the review comment of @uekerman https://github.com/precice/precice/pull/1386#discussion_r932446877

File: CMakeLists.txt (hunk 1 above)
Comment (fsimonis) [reply CMakeLists.txt:330]: This is merely to stay consistent with the `SolverInterface.hpp` file. I used the same reasoning for the `Tooling.hpp` file.

Our naming mixture is strange though.

File: docs/changelog/1392.md (hunk 2 above)
Comment (fsimonis) [reply docs/changelog/1392.md:1]: True. Probably even more relevant than the macros themselves.

File: src/precice/SolverInterface.hpp (hunk 3 above)
Comment (fsimonis) [reply src/precice/SolverInterface.hpp:7]: Including it by default means that you don't need to include the header in the adapter.
This makes your adapter forward compatible as the header doesn't exist in current releases.

**Comment by fsimonis** [comment]
> I'm not sure so let me ask: is the Version.h file installed in the target include directory or do we need to tell cmake?

That's part of CMake black magic. We declare PUBLIC_HEADERS in the source.cmake, which is generated by the updateSourceIndex.py script.
CMake installs these files when providing a destination for the PUBLIC_HEADERS in the call to install().

This is the easiest way until we have a separate include directory for interface files only.
Think:
```
.
├── cmake
│   └── modules
├── docs
├── examples
├── extras
├── includes [public include dir]
│   └── precice
│       ├── SolverInterface.hpp
│       ├── Version.h
│       └── Tooling.hpp
├── src [private include dir]
│   └── precice
│   │   ├── SolverInterface.cpp
│   │   ├── acceleration
│   │   ├── action
│   │   ├── com
│   │   ├── cplscheme
│   │   ├── drivers
│   │   ├── io
│   │   ├── impl
│   │   ├── logging
│   │   ├── m2n
│   │   ├── mapping
│   │   ├── math
│   │   ├── mesh
│   │   ├── partition
│   │   ├── query
│   │   ├── testing
│   │   ├── time
│   │   ├── utils
│   │   └── xml
├── tests
├── thirdparty
└── tools
```

:ghost: #263

**Pull request merged**